core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -4791,6 +4791,36 @@ object Stream extends StreamLowPriority {
     def lastOrError(implicit G: MonadError[G, Throwable]): G[O] =
       last.flatMap(_.fold(G.raiseError(new NoSuchElementException): G[O])(G.pure))
 
+    /** Compiles this stream in to a value of the target effect type `G`,
+      * raising a `NoSuchElementException` if the stream emitted no values, `IllegalStateException`
+      * if the stream emits more than one value and returning the only value emitted otherwise.
+      * Note that the stream execution will not short-circuit once it emits more than one element, it will still
+      * be evaluated to its end before raising the exception.
+      *
+      * When this method has returned, the stream has not begun execution -- this method simply
+      * compiles the stream down to the target effect type.
+      *
+      * @example {{{
+      * scala> import cats.effect.SyncIO
+      * scala> Stream(1).covary[SyncIO].compile.singletonOrError.unsafeRunSync()
+      * res0: Int = 1
+      * scala> Stream.empty.covaryAll[SyncIO, Int].compile.singletonOrError.attempt.unsafeRunSync()
+      * res1: Either[Throwable, Int] = Left(java.util.NoSuchElementException)
+      * scala> Stream.range(0,10).covary[SyncIO].compile.singletonOrError.attempt.unsafeRunSync()
+      * res2: Either[Throwable, Int] = Left(java.lang.IllegalStateException: Expected singleton stream)
+      * }}}
+      */
+    def singletonOrError(implicit G: MonadError[G, Throwable]): G[O] =
```

Comment:
Name is open for discussion. Doobie uses `unique` on queries, but I felt `singleton` being the more used term for streams.

Reply:
Thoughts on `.compile.only`?

Reply:
My first version was actually named `onlyOrError`. I would keep the `OrError` for consistency with `lastOrError` (and because there's a reasonable implementation of `only` returning `Option`), but other than that, happy to rename…

Reply:
OK, I prefer `onlyOnError` then :)